modules/hivemq-edge-module-opcua/src/main/java/com/hivemq/edge/adapters/opcua/OpcUaClientConnection.java

```diff
@@ -0,0 +1,350 @@
+            monitoredItem.setQueueSize(uint(config.getOpcuaToMqttConfig().getServerQueueSize()));
+            monitoredItem.setSamplingInterval(config.getOpcuaToMqttConfig().getPublishingInterval());
+            subscription.addMonitoredItem(monitoredItem);
+        });
+
+        try {
+            subscription.synchronizeMonitoredItems();
+            return CompletableFuture.completedFuture(null);
+        } catch (final MonitoredItemSynchronizationException ex) {
+            return CompletableFuture.failedFuture(ex);
+        }
+    }
+
+    private OpcUaSubscription.@NotNull SubscriptionListener createSubscriptionListener(
+            final @NotNull OpcUaClient client,
+            final @NotNull Map<NodeId, @NotNull OpcuaTag> nodeIdToTag,
+            final @NotNull Map<OpcuaTag, Boolean> tagToFirstSeen) {
+        return new OpcUaSubscription.SubscriptionListener() {
+            @Override
+            public void onKeepAliveReceived(final @NotNull OpcUaSubscription subscription) {
+                protocolAdapterMetricsService.increment(Constants.METRIC_SUBSCRIPTION_KEEPALIVE_COUNT);
+                OpcUaSubscription.SubscriptionListener.super.onKeepAliveReceived(subscription);
+            }
+
+            @Override
+            public void onTransferFailed(
+                    final @NotNull OpcUaSubscription subscription,
+                    final @NotNull StatusCode status) {
+                protocolAdapterMetricsService.increment(Constants.METRIC_SUBSCRIPTION_TRANSFER_FAILED_COUNT);
+                OpcUaSubscription.SubscriptionListener.super.onTransferFailed(subscription, status);
+            }
+
+            @Override
+            public void onDataReceived(
+                    final @NotNull OpcUaSubscription subscription,
+                    final @NotNull List<OpcUaMonitoredItem> items,
+                    final @NotNull List<DataValue> values) {
+                for (int i = 0; i < items.size(); i++) {
+                    final var tag = nodeIdToTag.get(items.get(i).getReadValueId().getNodeId());
+                    if (null == tagToFirstSeen.putIfAbsent(tag, true)) {
+                        eventService.createAdapterEvent(adapterId, PROTOCOL_ID_OPCUA)
+                                .withSeverity(Event.SEVERITY.INFO)
+                                .withMessage(String.format("Adapter '%s' took first sample for tag '%s'",
+                                        adapterId,
+                                        tag.getName()))
+                                .fire();
+                    }
+                    final var value = values.get(i);
+                    try {
+                        protocolAdapterMetricsService.increment(Constants.METRIC_SUBSCRIPTION_DATA_RECEIVED_COUNT);
+                        final var convertedPayload =
+                                new String(convertPayload(value, client.getDynamicEncodingContext()));
```

Comment:
So, the payload encoding is UTF-8? Are there any documents proving that? I suggest we leave a comment here.